### ISSUE TITLE
Invalid or unknown z ISA extension: 'zicbom'

### ISSUE THREAD

**Issue by nickdesaulniers** [ClangBuiltLinux/linux#1714]
@khilman [reports](https://github.com/kernelci/kernelci-core/pull/1447) that using BFD for riscv is failing with the error:
```
riscv64-linux-gnu-ld: -march=rv64i2p0_m2p0_a2p0_c2p0_zicbom1p0_zihintpause2p0: Invalid or unknown z ISA extension: 'zicbom'
```
example: https://storage.kernelci.org/next/master/next-20220920/riscv/defconfig+CONFIG_EFI=n/clang-16/logs/kernel.log

**Comment by nathanchance** [comment]
Isn't this on `ld.bfd` to support? Should we open an issue for it upstream? cc @palmer-dabbelt

**Comment by nickdesaulniers** [comment]
binutils-gdb commit 41d6ac5da655 ("RISC-V: Cache management instructions")
added support for Zicbom back in March 18.
Looking at the [releases](https://ftp.gnu.org/gnu/binutils/), it looks like that should have shipped in only the latest 2.39 release. KernelCI might be using an older release of GNU binutils. @khilman can you verify what version of riscv-linux-gnu-binutils KernelCI is using?

**Comment by khilman** [comment]
Looks like it's 2.35.2, which is the version of `binutils-riscv64-linux-gnu` that comes with debian bullseye, which is the baseline of our Docker images.
```
$ docker run -it kernelci/clang-16:kselftest-kernelci bash
kernelci@8f5aebbc8d3f:~$ riscv64-linux-gnu-ld --version
GNU ld (GNU Binutils for Debian) 2.35.2
Copyright (C) 2020 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) a later version.
This program has absolutely no warranty.
kernelci@8f5aebbc8d3f:~$ clang --version
Debian clang version 16.0.0-++20220919071801+818b1ab84e7b-1~exp1~20220919071900.393
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/lib/llvm-16/bin
kernelci@8f5aebbc8d3f:~$ 
```

**Comment by ConchuOD** [comment]
Sounds about right. @nathanchance if you're thinking about combos to put into Palmer's pre-commit CI then prob some sort of llvm/binutils combo seem like a good idea.

**Comment by nathanchance** [comment]
> Sounds about right. @nathanchance if you're thinking about combos to put into Palmer's pre-commit CI then prob some sort of llvm/binutils combo seem like a good idea.

Yup, I agree! I am not sure that it would have really saved us here though, as we will likely be using the latest stable version of binutils so we would have support for `zicbom`. I am not sure how Palmer has the versioning set up yet though, I'll be studying it soon.

**Comment by ConchuOD** [comment]
> > Sounds about right. @nathanchance if you're thinking about combos to put into Palmer's pre-commit CI then prob some sort of llvm/binutils combo seem like a good idea.
> 
> Yup, I agree! I am not sure that it would have really saved us here though, as we will likely be using the latest stable version of binutils so we would have support for `zicbom`. I am not sure how Palmer has the versioning set up yet though, I'll be studying it soon.

Yeah, less about this specific issue and more about me suspecting that neither developers nor reviewers are using a mixed setup.

**Comment by khilman** [comment]
For KernelCI, I'm now removing the mixed setu.  Seems like it was added as a workaround, but was then forgotten.

**Comment by ConchuOD** [comment]
FYI this is *not* an interop bug, it's broken for gcc 12 + binutils w/o zicbom too.

**Comment by ConchuOD** [comment]
> Isn't this on `ld.bfd` to support? Should we open an issue for it upstream? cc @palmer-dabbelt

I had another bit of a look at this tonight, LD_VERSION >= 2.39 (LLD_VERSION would be 15) is required - but gating the Zicbom symbols based on that doesn't stop the string being emitted into the object file, so not sure how one is meant to handle that.

**Comment by palmer-dabbelt** [comment]
oops, I guess we need to check `-Wa,-march=...zicbom...` rather than just `-mach=...zicbom...`, because that's what we need to make inline assembly work.  Does someone have a patch?

**Comment by ConchuOD** [comment]
if it's that straightforward, I guess I can have one...

**Comment by ConchuOD** [comment]
Seemingly not that straightforward. That does do dynamic disabling of CONFIG_...ZICBOM but it does not stop zicbom being emitted into the object files.

**Comment by ConchuOD** [comment]
> Seemingly not that straightforward. That does do dynamic disabling of CONFIG_...ZICBOM but it does not stop zicbom being emitted into the object files.

Palmer pointed out I missed a cc-option check in the Makefile too... That sorts it so I will in fact send a aptch...

**Comment by ConchuOD** [comment]
Ignore those commits, I had a duplicate as-option check in them - but I submitted some patches that seem to work..

I've probably done something awful since tc stuff is pushing my comfort zone, but:
https://lore.kernel.org/linux-riscv/20221006173520.1785507-1-conor@kernel.org

**Comment by ConchuOD** [comment]
https://git.kernel.org/pub/scm/linux/kernel/git/palmer/linux.git/commit/?h=fixes&id=952b64d6665c8b6fdd13ba63be7adc3be51641d7

**Comment by nathanchance** [comment]
https://git.kernel.org/linus/952b64d6665c8b6fdd13ba63be7adc3be51641d7
